syntax = "proto3";

package qbrix.proxy;

import "common.proto";

service ProxyService {
  rpc CreatePool(CreatePoolRequest) returns (CreatePoolResponse);
  rpc GetPool(GetPoolRequest) returns (GetPoolResponse);
  rpc DeletePool(DeletePoolRequest) returns (DeletePoolResponse);

  rpc CreateExperiment(CreateExperimentRequest) returns (CreateExperimentResponse);
  rpc GetExperiment(GetExperimentRequest) returns (GetExperimentResponse);
  rpc UpdateExperiment(UpdateExperimentRequest) returns (UpdateExperimentResponse);
  rpc DeleteExperiment(DeleteExperimentRequest) returns (DeleteExperimentResponse);

  rpc CreateGateConfig(CreateGateConfigRequest) returns (CreateGateConfigResponse);
  rpc GetGateConfig(GetGateConfigRequest) returns (GetGateConfigResponse);
  rpc UpdateGateConfig(UpdateGateConfigRequest) returns (UpdateGateConfigResponse);
  rpc DeleteGateConfig(DeleteGateConfigRequest) returns (DeleteGateConfigResponse);

  rpc Select(SelectRequest) returns (SelectResponse);
  rpc Feedback(FeedbackRequest) returns (FeedbackResponse);
  rpc Health(qbrix.common.HealthCheckRequest) returns (qbrix.common.HealthCheckResponse);
}

// Pool management
message CreatePoolRequest {
  string name = 1;
  repeated qbrix.common.Arm arms = 2;
}

message CreatePoolResponse {
  qbrix.common.Pool pool = 1;
}

message GetPoolRequest {
  string pool_id = 1;
}

message GetPoolResponse {
  qbrix.common.Pool pool = 1;
}

message DeletePoolRequest {
  string pool_id = 1;
}

message DeletePoolResponse {
  bool deleted = 1;
}

// Experiment management
message CreateExperimentRequest {
  string name = 1;
  string pool_id = 2;
  string protocol = 3;
  map<string, string> protocol_params = 4;
  bool enabled = 5;

  // Feature gate config
  FeatureGateConfig feature_gate = 6;
}

message FeatureGateConfig {
  bool enabled = 1;
  double rollout_percentage = 2;  // 0.0 - 100.0
  string default_arm_id = 3;      // arm to return when gated
  ScheduleConfig schedule = 4;
  ActiveHoursConfig active_hours = 5;
  string timezone = 6;            // e.g., "UTC", "America/New_York"
  repeated RuleConfig rules = 7;
}

message ScheduleConfig {
  optional int64 start_timestamp_ms = 1;
  optional int64 end_timestamp_ms = 2;
}

message ActiveHoursConfig {
  optional string start = 1;  // HH:MM format
  optional string end = 2;    // HH:MM format
}

message RuleConfig {
  string key = 1;             // metadata key to evaluate
  string operator = 2;        // ==, !=, >, <, contains, in, etc.
  string value = 3;           // value to compare against (JSON encoded for complex types)
  string arm_id = 4;          // arm to return if rule matches
}

message CreateExperimentResponse {
  qbrix.common.Experiment experiment = 1;
}

message GetExperimentRequest {
  string experiment_id = 1;
}

message GetExperimentResponse {
  qbrix.common.Experiment experiment = 1;
  FeatureGateConfig feature_gate = 2;
}

message UpdateExperimentRequest {
  string experiment_id = 1;
  optional bool enabled = 2;
  map<string, string> protocol_params = 3;
  optional FeatureGateConfig feature_gate = 4;
}

message UpdateExperimentResponse {
  qbrix.common.Experiment experiment = 1;
}

message DeleteExperimentRequest {
  string experiment_id = 1;
}

message DeleteExperimentResponse {
  bool deleted = 1;
}

// Gate config management
message CreateGateConfigRequest {
  string experiment_id = 1;
  FeatureGateConfig config = 2;
}

message CreateGateConfigResponse {
  FeatureGateConfig config = 1;
}

message GetGateConfigRequest {
  string experiment_id = 1;
}

message GetGateConfigResponse {
  FeatureGateConfig config = 1;
}

message UpdateGateConfigRequest {
  string experiment_id = 1;
  FeatureGateConfig config = 2;
}

message UpdateGateConfigResponse {
  FeatureGateConfig config = 1;
}

message DeleteGateConfigRequest {
  string experiment_id = 1;
}

message DeleteGateConfigResponse {
  bool deleted = 1;
}

// Selection (external API)
message SelectRequest {
  string experiment_id = 1;
  qbrix.common.Context context = 2;
}

message SelectResponse {
  qbrix.common.Arm arm = 1;
  string request_id = 2;
  bool is_default = 3;  // True if returned by feature gate, not bandit
}

// Feedback
message FeedbackRequest {
  string request_id = 1;  // signed token from SelectResponse containing all selection context
  double reward = 2;
}

message FeedbackResponse {
  bool accepted = 1;
}